### ISSUE TITLE
damping method and matrix solver

### ISSUE THREAD

**Issue by raphaelvalentin** [fabiodimarco/tf-levenberg-marquardt#13]
Hi Fabio, 

1. **damping method**

Based on two interesting documents:
[1] http://www2.imm.dtu.dk/pubdb/edoc/imm3215.pdf
[2] https://people.duke.edu/~hpgavin/ce281/lm.pdf

It seems that it has been proposed that the LM damping method features a normalization to JJT (`max(diag(JJT)`) that may provide better and more stable minimization.

The code is in `DampingAlgorithm`

```
            damping = tf.eye(tf.shape(JJ)[0], dtype=JJ.dtype)

        damping = tf.scalar_mul(damping_factor, damping)
        return tf.add(JJ, damping)
```

Following the documents, I propose this modification:

```
            max_diag = tf.math.reduce_max(tf.linalg.diag_part(JJ))
            damping = max_diag * tf.eye(tf.shape(JJ)[0], dtype=JJ.dtype)

        damping = tf.scalar_mul(damping_factor, damping)
        return tf.add(JJ, damping)
```

with a `starting_value=1e-6`  and `tf.random.set_seed(1234)`, I can get these results from your provided example.

![Figure_3](https://user-images.githubusercontent.com/12194328/175805562-37fe2d43-23c5-42b6-baf3-31a7d3ec9de0.png)

We can observe clearly that the minimization is more stable; indeed, at same number of steps, we can observe a decade smaller in loss. Of course, it may require more tests to see if this factor helps in all cases. Also, considering the value of the loss, I may suggest to set the property `tf.keras.backend.set_floatx('float64')` for the curve fitting.

In addition, the document [1] (pp.8) shows that it is possible to observe a benefit when the damped method is monitored by a step length. It shows particularly that the _Nielsen_ strategy may help on the general stability of the optimizer.

2. **Cholesky method**

The Cholesky method is probably an efficient numerical solutions to solve a square matrix. Unfortunately, by definition, Cholesky requires a positive definite matrix (due to the square root). However, it may have a simple way to reach this requirement with LM which is described in Appendix A of document [1].

As instance, this code, written with `numpy` (not `tensorflow`) shows that we can get a positive definite matrix by just increasing the LM damping parameter:

```
        while 1:
            try:
                # if not positive definite, we damp mu
                L = np.cholesky(self.A + self.mu*I)
                y = np.lstsq(L, self.g.T[0], rcond=None)[0]
                self.dX = np.lstsq(L.T, y, rcond=None)[0]
            except LinAlgError as e:
                self.mu = 10*self.mu
            else:
                break
```

Of course, this piece of code needs to be converted using methods of `tensorflow`. 
There are certain advantages by using Cholesky method. We can keep the convergence of the optimizer (as QR) and the speed can be improved sensitively (faster by a factor of ~3). 
Moreover, when matrix is not well formed (near-singular and rank deficient matrices), it mostly has consequence that the resulted LM step will not be so good; an increase of the damping will provide a better step. Then, it may help to improve the stability of the optimizer.

What do you think about this?

** I apologize to have put 2 different topics on same issue thread. If you like, I can copy paste in another thread.

Thanks for your support, 
Raphael.

**Comment by fabiodimarco** [comment]
Hi Raphael,

thank you for the tips, I think I can add the damping normalization with `max(diag(JJT)` as another option to the DampingAlgorithm class in addition to the fletcher method already available. I'll try it on the two examples I provided to see if it improves the convergence speed.

I agree on using `tf.keras.backend.set_floatx('float64')` for curve fitting applications, it can be useful and I have used it several times for my experiments. In the examples I used float32, as it is the most common format in machine learning applications, where reducing the loss as much as possible is not the main purpose.

The solution you mentioned of increasing the damping when the matrix is ​​singular is very useful not only for the Cholesky method, but also when using the QR method. Sometimes I had to increase the min_value of the damping which is `1e-10` by default, because it gave me runtime errors even with QR method.
However, I tried to implement it some time ago, but was not able to make it work in Tensorflow. However, I think the problems I was having on try except were due to the fact that I was using `@tf.function` on every function of the code. In the current version, however, the code uses `@tf.function` only for the Jacobian computation (since using it everywhere did not provide performance improvements and was even worse in some cases).
I'll have to do some experiment, to see if I can get it to work using try except similar to the example you provided.

Best,
Fabio

**Comment by raphaelvalentin** [comment]
Hi Fabio, 

Documents that catch my interest quite a lot : 
[1] [zhang2017.pdf](https://github.com/fabiodimarco/tf-levenberg-marquardt/files/8987757/zhang2017.pdf)
[2] 
[Physics-Inspired_Neural_Networks_Pi-NN_for_Effcien.pdf](https://github.com/fabiodimarco/tf-levenberg-marquardt/files/8987883/Physics-Inspired_Neural_Networks_Pi-NN_for_Effcien.pdf)
[3] https://curve.carleton.ca/system/files/etd/3f17140b-feb5-4b8c-bc08-0d7217977145/etd_pdf/c3a0a1172423ed7032a74896b52b411c/sadrossadat-sensitivityanalysisbasedadjointneuralnetwork.pdf
[4] [Modeling_of_VLSI_MOSFET_Characteristics_Using_Neur.pdf](https://github.com/fabiodimarco/tf-levenberg-marquardt/files/8987816/Modeling_of_VLSI_MOSFET_Characteristics_Using_Neur.pdf)
[5] [NeuroFET_Webcast_Final.pdf](https://github.com/fabiodimarco/tf-levenberg-marquardt/files/8987823/NeuroFET_Webcast_Final.pdf)

LM class optimizer can offer clear opportunities in term of accuracy and speed. Based on my experiments, your code implementation that you provide is really very interesting for such case.

Sincerely, 
Raphael.

**Comment by fabiodimarco** [comment]
Hi Raphael,

I updated the code to add the changes we discussed.
1) The DampingAlgorithm now have a adaptive_scaling option (True/False).
2) The train_step now handles exceptions in case the decomposition / solve fails.

The only "problem" encountered is that for tensorflow 2.4.0 in case of singular matrices the Cholesky decomposition throws an exception (as expected), this does not happen in tensorflow 2.9.0 where the output matrix is filled with NaN and this error message occurs.
`Cholesky decomposition was not successful. Eigen::LLT failed with error code 1. Filling lower-triangular output with NaNs.`

However the code currently handles both cases correctly, although I believe this can be considered a bug of the new version of Tensorflow.

If you have time let me know if it works as you expected.

Best,
Fabio
